Kernel/LPOComparator.cpp

```diff
@@ -0,0 +1,491 @@
+  int prevIndex = INDEX_UNINITIALIZED;
+  for (unsigned j = i; j < t->arity(); j++) {
+    auto compRes = createHelper(tl1,*t->nthArgument(j),lpo);
+    // If the comparison is 'greater', we skip this iteration.
+    if (compRes->second == BranchTag::T_GREATER) {
+      continue;
+    }
+    // Otherwise, if the comparison is 'equal' or 'incomparable',
+    // we can return early with a negative result.
+    if (compRes->second != BranchTag::T_JUMP) {
+      res.reset();
+      return make_pair(res,BranchTag::T_INCOMPARABLE);
+    }
+    // Update previous 'greater' branches
+    // to a jump to the current index.
+    if (prevIndex != INDEX_UNINITIALIZED) {
+      updateBranchInRange(res, prevIndex, (unsigned)res.size(),
+        Branch::gt(), Branch::jump(res.size()));
+    }
+    prevIndex = res.size();
+    // Push the comparison instructions and replace each
+    // 'equal' with 'incomparable' in them.
+    pushInstructions(res, compRes->first, Branch::inc(), Branch::gt(), Branch::inc());
+  }
+  return make_pair(res,res.isEmpty() ? BranchTag::T_GREATER : BranchTag::T_JUMP);
+}
+
+/**
+ * Implements an @b LPO::alpha call via instructions.
+ */
+pair<Stack<Instruction>,BranchTag> LPOComparator::alphaChain(const LPO& lpo, Term* s, unsigned i, TermList tl2)
+{
+  Stack<Instruction> res;
+  int prevIndex = INDEX_UNINITIALIZED;
+  for (unsigned j = i; j < s->arity(); j++) {
+    auto compRes = createHelper(*s->nthArgument(j),tl2,lpo);
+    // If the comparison is 'incomparable', we skip this iteration.
+    if (compRes->second == BranchTag::T_INCOMPARABLE) {
+      continue;
+    }
+    // Otherwise, if the comparison is 'greater' or 'equal',
+    // we can return early with a positive result.
+    if (compRes->second != BranchTag::T_JUMP) {
+      res.reset();
+      return make_pair(res,BranchTag::T_GREATER);
+    }
+    // Update previous 'incomparable' branches
+    // to a jump to the current index.
+    if (prevIndex != INDEX_UNINITIALIZED) {
+      updateBranchInRange(res, prevIndex, (unsigned)res.size(),
+        Branch::inc(), Branch::jump(res.size()));
+    }
+    prevIndex = res.size();
+    // Push the comparison instructions and replace each
+    // 'equal' with 'greater' in them.
+    pushInstructions(res, std::move(compRes->first), Branch::gt(), Branch::gt(), Branch::inc());
+  }
+  return make_pair(res,res.isEmpty() ? BranchTag::T_INCOMPARABLE : BranchTag::T_JUMP);
+}
+
+pair<Stack<Instruction>,BranchTag>* LPOComparator::createHelper(TermList tl1, TermList tl2, const LPO& lpo)
+{
+  static DHMap<std::pair<TermList,TermList>,pair<Stack<Instruction>,Instruction::BranchTag>*> _cache;
+
+  pair<Stack<Instruction>,BranchTag>** ptr;
+  // We have a local cache for subresults to avoid too much computation.
+  if (!_cache.getValuePtr(make_pair(tl1,tl2),ptr)) {
+    return *ptr;
+  }
+  // Allocate on heap so that cache reallocation
+  // won't affect partial results.
+  auto res = new pair(Stack<Instruction>(), BranchTag::T_JUMP);
```

Comment:
This is probably quite hot and creates a double-indirection which it might be possible to avoid - for now I think you are ironing out performance later.

Reply:
Yes, this is stupid. I think we could just move a temporary result into the cache at the very end.

Reply:
Turns out the main problem is that handle multiple subresults as pointers, and these could change during reallocation. An alternative evil is to copy the subresults to temporary objects which seems even less efficient. What do you think?

Reply:
In a future version of this I might have (i) a hashmap to indices and (ii) an array of these objects - this is still an indirection, of course, but it's a bit more spatial and one can't forget to delete things.

Or, it seems from a quick skim that the results are "finished" and their stacks of instructions won't change after `createHelper` returns. If this is true, we could return a `Instruction *` - length - `BranchTag` triple from `createHelper` and have the cache alone own the `const Stack<Instruction>`.

But for now this works!